legion/engine/core/particles/particlesystem.cpp

```diff
@@ -0,0 +1,117 @@
+#include <core/particles/particles.hpp>
+#include <core/common/debugrendering.hpp>
+
+
+namespace legion::core
+{
+    void ParticleSystem::setup()
+    {
+        log::filter(log::severity_debug);
+        log::debug("ParticleSystem setup");
+
+        bindToEvent<events::component_creation<particle_emitter>, &ParticleSystem::emitter_setup>();
+    }
+
+    void ParticleSystem::shutdown()
+    {
+        log::debug("ParticleSystem shutdown");
+    }
+
+    void ParticleSystem::update(time::span deltaTime)
+    {
+        ecs::filter<particle_emitter> filter;
+        for (auto& ent : filter)
+        {
+            auto& emitter = ent.get_component<particle_emitter>().get();
+            if (emitter.currentParticleCount < emitter.maxSpawnCount)
+            {
+                emitter.elapsedTime += deltaTime;
+                if (emitter.elapsedTime >= emitter.spawnInterval)
+                {
+                    emit(emitter, emitter.spawnRate);
+                    emitter.elapsedTime = 0;
+                }
+            }
+            maintanence(emitter, deltaTime);
+        }
+    }
+
+    void ParticleSystem::emitter_setup(L_MAYBEUNUSED events::component_creation<particle_emitter>& event)
+    {
+        ecs::filter<particle_emitter> filter;
+        for (auto& ent : filter)
+        {
+            auto emitter = ent.get_component<particle_emitter>();
+            for (size_type i = 0; i < emitter->maxSpawnCount; i++)
+                emitter->set_alive(i, false);
+        }
+    }
+
+    void ParticleSystem::emit(particle_emitter& emitter, size_type count)
+    {
+        auto& ageBuffer = emitter.get_buffer<life_time>("lifetimeBuffer");
+        if (emitter.currentParticleCount + count > emitter.maxSpawnCount)
+            count = emitter.maxSpawnCount - emitter.currentParticleCount;
+
+        for (size_type i = 0; i < count; i++)
+        {
+            size_type targetIdx = emitter.currentParticleCount + i;
+            emitter.set_alive(targetIdx, true);
+            ageBuffer.get(targetIdx).age = 0;
+            ageBuffer.get(targetIdx).max = math::linearRand(emitter.minLifeTime, emitter.maxLifeTime);
+        }
+
+        for (auto& policy : emitter.particlePolicies)
+            policy->OnInit(emitter, emitter.currentParticleCount, emitter.currentParticleCount + count);
+
+        emitter.currentParticleCount += count;
+    }
```

Comment:
the compiler might optimize `emitter.currentParticleCount + count` out, but it won't do that when optimization is turned off (debug mode) so then you might only wanna calculate that addition once and reuse it